WordPress/gutenberg
# E2E: Fix artifacts handling in CI

## What?

Resolves #58995.
Supersedes #57183.

Since the [artifacts became immutable in v4 of actions/upload-artifact](https://github.com/actions/upload-artifact#:~:text=The%20contents%20of%20an%20Artifact%20are%20uploaded%20together%20into%20an%20immutable%20archive.%20They%20cannot%20be%20altered%20by%20subsequent%20jobs%20unless%20the%20Artifacts%20are%20deleted%20and%20recreated%20(where%20they%20will%20have%20a%20new%20ID).%20Both%20of%20these%20factors%20help%20reduce%20the%20possibility%20of%20accidentally%20corrupting%20Artifact%20files.), we must create a separate artifact for each shard (both failures and flaky reports), or we'll get a conflict error (see the attached issue #58995). This is the main issue this PR addresses.

#### Enhancements

For convenience, this PR adds a step to merge artifacts into a single zip so we don't need to remember the shard number. The partial artifacts will be deleted after merging to reduce noise. In case of a rerun, the new artifacts (if any) will be merged with those from the previous run(s) so we don't lose any information. Lastly, unlike currently, the `Report to GitHub` job will only run if a flaky test report zip is produced.

## Testing Instructions

Since this is not manually testable, I've pushed a [tmp commit](https://github.com/WordPress/gutenberg/pull/61338/commits/5e27003bab65914d6ecff335f96fbc199b8ab4f2) that causes two tests to be flaky. Those tests are in two shards (no. 1 & 2) to verify that the artifact upload, merging, and flaky test reporting work as expected. Download the artifacts and confirm the content matches what's reported in CI. Check out the CI output and ensure everything looks as expected. Confirm the [flaky report comment](https://github.com/WordPress/gutenberg/pull/61338#issuecomment-2094346787) contains correct test reports.

_The tmp commit will be reverted when this PR is approved. The reopened flaky report issues will also be closed._
